### DISCUSSION TITLE
Tasmota web interface need improvement if more than 4 or 8 relays are present.

### DISCUSSION THREAD

**Discussion by rapi3** [arendst/Tasmota#11546] [Ideas]
Hello,

I am experimenting with MCP23017 and I notice some things that are wrong and need a soft touch:
at first look you may not notice that state it is not ok, but it is shifted to right:
for ...7, 8... actual state it displayed over ... 9, 10...
![mcp23017-1](https://user-images.githubusercontent.com/18531150/113345791-3d89fa80-9333-11eb-8e02-a4fed2def01c.png)
...
It will be great if states can be displayed on 2 or more rows if more than 8 relays are present,
maybe someone know where do I need to look to try to change how it is displayed for example like this:
![mcp23017-8](https://user-images.githubusercontent.com/18531150/113347669-cbff7b80-9335-11eb-8ed7-8eb5d006b4e2.png)

_Or even better to have the same look as RFBridge if more than 4 relays are present but with a small or big change:
to have the red button if state it is 1 and blue if it is 0:_
![RFB-1-red](https://user-images.githubusercontent.com/18531150/113348701-395fdc00-9337-11eb-80f6-d9ed30e1ad2d.png)

Do you think it is possible ?

**Comment by arendst** [comment]
I don't see your issue:

![image](https://user-images.githubusercontent.com/11044339/113405589-600b2a80-93aa-11eb-8c48-caf3e097b45b.png)

What browser do you use? How come the default font is Times Roman?

**Comment by arendst** [comment]
Ah I see. It's that damn Firefox again:

![image](https://user-images.githubusercontent.com/11044339/113405828-c09a6780-93aa-11eb-97d8-19e0853fadc1.png)

Time to move to chrome guys!

**Comment by rapi3** [comment]
Am I wrong if I think that using buttons with colors for state change ( red for ON / 1 ) and discard text over buttons will save some space in flash/memory ?

**_Maybe it can be available at compile at user define._**

p.s.
yes I am using SeaMonkey/Firefox on Linux because I can share the same settings and bookmarks folder with windows

**Comment by arendst** [comment]
You are probably wrong ;-) The current implementation updates the state of the switches regularly while keeping the information on the buttons stable. A change to color would mean the buttons will need to be rewritten regularly instead.

Having a compile option is nice. Just find one who will want to rewrite the current GUI to your likings while keeping the code footprint (and server/client datatransfer) low.

Another point is that I seem to be only person in this world who never understand the meaning of a color with regards of a state; is red Off or On and reverse. I hate these slide buttons too as I never know in what state they are.

**Comment by rapi3** [comment]
Thank you for clarification.

What about splitting buttons on two rows if more than 8 relay are available ( as in picture 2 ) is that a lot of code to change ?
That will help when use the mobile phone.

**Comment by arendst** [comment]
I'll chk if it can be done easily and still makes sense.

**Comment by rapi3** [comment]
thank you very much :+1:

**Comment by arendst** [comment]
Give the latest development a try:

![image](https://user-images.githubusercontent.com/11044339/113420017-dc146b00-93c8-11eb-8cbd-6d38e009b27b.png)

**Comment by rapi3** [comment]
Many thanks,
interface it looks very good and it is much easy to follow the relays now:

![RFID-1](https://user-images.githubusercontent.com/18531150/113422996-00bf1180-93ce-11eb-8350-c0ea645a2f3f.png)

p.s.
Id like to try to have Configuration swapped with Firmware Upgrade because already few time I made the mistake to click on button toggle on a 1000km remote plug that switched off the wifi AP ( lucky me I had the rule to turn back on 3 min after power off ) and now I use PowerOnState 4 to prevent this mistake but this prevent local Power OFF/ON/Reset from button if required.

Can you point me where in the code it is the order of menu displayed:
Configuration
Information
Firmware Upgrade
Console
Restart

**p.p.s**
There is a problem if Pulsetime it is activated for relays 1-8; it will interfere with relays 9-16
and on console for relays 9-16 pulsetime is unavailable
15:38:41.212 CMD: pulsetime9
15:38:41.217 MQT: Wemos/stat/W11/RESULT = {"Command":"Error"}

_It si easy to reproduce:_
Activate PulseTime for one or more relays 1-8 and you can see his effect on correspondent relay 9-16 when you click correspondent relay.

**Reply by arendst** [reply]
> Can you point me where in the code it is the order of menu displayed:

Play around with this in xdrv_01_webserver.ino
```  if (HTTP_ADMIN == Web.state) {
#ifdef FIRMWARE_MINIMAL
    WSContentSpaceButton(BUTTON_FIRMWARE_UPGRADE);
#else
    WSContentSpaceButton(BUTTON_CONFIGURATION);
    WSContentButton(BUTTON_INFORMATION);
    WSContentButton(BUTTON_FIRMWARE_UPGRADE);
#endif  // Not FIRMWARE_MINIMAL
    WSContentButton(BUTTON_CONSOLE);
    WSContentButton(BUTTON_RESTART);
  }
```

> There is a problem if Pulsetime it is activated for relays 1-8; it will interfere with relays 9-16

As designed due to memory constraints. It's documented somewhere.

**Comment by rapi3** [comment]
Super many thanks for help :+1:

**Comment by rapi3** [comment]
OK,
I succeed to do that and now I need your help again to move the space between Information and Configuration where it belong up must be hard coded in Configuration button somewhere ?

![RFID-2](https://user-images.githubusercontent.com/18531150/113430133-ef7c0200-93d9-11eb-81b0-4e5fb9a7bebb.png)
.

**Reply by arendst** [reply]
Check the function names and play with it.

**Comment by rapi3** [comment]
Once again thank you for your help :+1: 
In my defense I have to confess now that my C and compiling knowledge it is 0+ but I succeed to do it.

![RFID-3](https://user-images.githubusercontent.com/18531150/113432179-34edfe80-93dd-11eb-81be-a66ef34af1cd.png)
